cpu/gd32v/periph/uart.c

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright (C) 2020 Koen Zandberg <koen@bergzand.net>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+/**
+ * @ingroup         cpu_gd32v
+ * @{
+ *
+ * @file
+ * @brief           GD32V UART peripheral implementation
+ *
+ * @author          Koen Zandberg <koen@bergzand.net>
+ */
+#include <assert.h>
+#include <stdint.h>
+
+#include "periph/uart.h"
+#include "periph/gpio.h"
+#include "periph_cpu.h"
+#include "cpu.h"
+#include "clic.h"
+
+#define RXENABLE            (USART_CTL0_REN_Msk | USART_CTL0_RBNEIE_Msk)
+
+/**
+ * @brief   Allocate memory to store the callback functions
+ *
+ * Extend standard uart_isr_ctx_t with data_mask field. This is needed
+ * in order to mask parity bit.
+ */
+static struct {
+    uart_rx_cb_t rx_cb;   /**< data received interrupt callback */
+    void *arg;            /**< argument to both callback routines */
+    uint8_t data_mask;    /**< mask applied to the data register */
+} isr_ctx[UART_NUMOF];
+
+
+static inline USART_Type *dev(uart_t uart)
+{
+    return uart_config[uart].dev;
+}
+
+static inline void uart_init_pins(uart_t uart, uart_rx_cb_t rx_cb)
+{
+     /* configure TX pin */
+    gpio_init_af(uart_config[uart].tx_pin, GPIO_AF_OUT_PP);
+    /* configure RX pin */
+    if (rx_cb) {
+        gpio_init(uart_config[uart].rx_pin, GPIO_IN_PU);
+    }
+}
+
+static inline void uart_enable_clock(uart_t uart)
+{
+    /* TODO: add pm blocker */
+    periph_clk_en(uart_config[uart].bus, uart_config[uart].rcu_mask);
+}
+
+static inline void uart_disable_clock(uart_t uart)
+{
+    periph_clk_dis(uart_config[uart].bus, uart_config[uart].rcu_mask);
+    /* TODO remove pm blocker */
+}
+
+static inline void uart_init_usart(uart_t uart, uint32_t baudrate)
+{
+    uint16_t mantissa;
```

Comment:
mantissa? Is this to match the terminology of the datasheet? If so, maybe add `/* (sic!) */`?

Reply:
Copied from the compatible [stm32 implementation](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32/periph/uart.c#L295-L297) :)

Reply:
According to [this article on hackaday](https://hackaday.com/2021/01/08/bare-metal-stm32-universal-asynchronous-communication-with-uarts/), the ST guys confused remainder with mantissa:

> Without going into too much detail (for that see e.g. 27.5.4 in RM0091), the simple way to fill in this register without changing the default settings (like enabling OVER8 in USART_CR1), is to divide the system core clock by the desired baudrate, then dividing this by 16 twice, first to get the integer fraction and then using a modulo operation to get the remainder (incorrectly called the ‘mantissa’ by ST).

Care to add the `/* (sic!) /*` to make it obvious that the variable named `mantissa` doesn't really store a mantissa? Otherwise this confuses the shit out of me.

Reply:
How about I just rename the variable to something not confusing? Suggestions welcome :)